src/algebra/direct_limit.lean

```diff
@@ -201,47 +201,41 @@ variables [Π i, add_comm_group (G i)]
 include dec_ι
 
 /-- The direct limit of a directed system is the abelian groups glued together along the maps. -/
-def direct_limit (f : Π i j, i ≤ j → G i → G j) [Π i j hij, is_add_group_hom (f i j hij)] : Type* :=
+def direct_limit (f : Π i j, i ≤ j → G i →+ G j) : Type* :=
 @module.direct_limit ℤ _ ι _ _ G _ _
-  (λ i j hij, (add_monoid_hom.of $ f i j hij).to_int_linear_map)
+  (λ i j hij, (f i j hij).to_int_linear_map)
 
 namespace direct_limit
 
-variables (f : Π i j, i ≤ j → G i → G j)
-variables [Π i j hij, is_add_group_hom (f i j hij)]
+variables (f : Π i j, i ≤ j → G i →+ G j)
 
 omit dec_ι
 
-protected lemma directed_system [directed_system G f] :
-  module.directed_system G (λ i j hij, (add_monoid_hom.of $ f i j hij).to_int_linear_map) :=
-⟨directed_system.map_self f, directed_system.map_map f⟩
+protected lemma directed_system [directed_system G (λ i j h, f i j h)] :
+  module.directed_system G (λ i j hij, (f i j hij).to_int_linear_map) :=
+⟨directed_system.map_self (λ i j h, f i j h), directed_system.map_map (λ i j h, f i j h)⟩
 
 include dec_ι
 
 local attribute [instance] direct_limit.directed_system
 
 instance : add_comm_group (direct_limit G f) :=
-module.direct_limit.add_comm_group G (λ i j hij, (add_monoid_hom.of $f i j hij).to_int_linear_map)
+module.direct_limit.add_comm_group G (λ i j hij, (f i j hij).to_int_linear_map)
 
 instance : inhabited (direct_limit G f) := ⟨0⟩
 
 /-- The canonical map from a component to the direct limit. -/
-def of (i) : G i → direct_limit G f :=
-module.direct_limit.of ℤ ι G (λ i j hij, (add_monoid_hom.of $ f i j hij).to_int_linear_map) i
+def of (i) : G i →ₗ[ℤ] direct_limit G f :=
+module.direct_limit.of ℤ ι G (λ i j hij, (f i j hij).to_int_linear_map) i
 variables {G f}
 
-instance of.is_add_group_hom (i) : is_add_group_hom (of G f i) :=
-linear_map.is_add_group_hom _
-
 @[simp] lemma of_f {i j} (hij) (x) : of G f j (f i j hij x) = of G f i x :=
 module.direct_limit.of_f
 
-@[simp] lemma of_zero (i) : of G f i 0 = 0 := is_add_group_hom.map_zero _
-@[simp] lemma of_add (i x y) : of G f i (x + y) = of G f i x + of G f i y :=
-is_add_hom.map_add _ _ _
-@[simp] lemma of_neg (i x) : of G f i (-x) = -of G f i x := is_add_group_hom.map_neg _ _
-@[simp] lemma of_sub (i x y) : of G f i (x - y) = of G f i x - of G f i y :=
-is_add_group_hom.map_sub _ _ _
+@[simp] lemma of_zero (i) : of G f i 0 = 0 := linear_map.map_zero _
+@[simp] lemma of_add (i x y) : of G f i (x + y) = of G f i x + of G f i y := linear_map.map_add _ _ _
+@[simp] lemma of_neg (i x) : of G f i (-x) = -of G f i x := linear_map.map_neg _ _
+@[simp] lemma of_sub (i x y) : of G f i (x - y) = of G f i x - of G f i y := linear_map.map_sub _ _ _
```

Comment:
We can delete these because `linear_map.map_` already covers them, right?

Reply:
I'm not sure - wouldn't the simp linter complain if these were "already covered"? Aside from that, it's still easier to write `rw of_neg` than it is to write `rw (of _ _ _).map_neg`,

Reply:
I believe the `simp` linter doesn't notice it when you define a (redundant) specialization after you define the original, because `simp` tries to apply later definitions before applying earlier definitions.

Reply:
As for why I prefer to remove the lemmas: it's less code to maintain, which is always good. And by making `(of _ _ _).map_neg` (or `linear_map.map_neg`) the only one way to do it, it ensures our usage is less surprising: easier to understand.

The precedent is also to remove these extras, it seems: see #4011.

Reply:
Thanks for the link for precedent, that's fairly compelling. I've remove all the lemmas.